crates/iceberg/src/arrow/reader.rs

```diff
@@ -186,4 +221,637 @@ impl ArrowReader {
+    }
+
+    fn greater_than_or_eq(
+        &mut self,
+        reference: &BoundReference,
+        _literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn eq(
+        &mut self,
+        reference: &BoundReference,
+        _literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn not_eq(
+        &mut self,
+        reference: &BoundReference,
+        _literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn starts_with(
+        &mut self,
+        reference: &BoundReference,
+        _literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn not_starts_with(
+        &mut self,
+        reference: &BoundReference,
+        _literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn r#in(
+        &mut self,
+        reference: &BoundReference,
+        _literals: &FnvHashSet<Datum>,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+
+    fn not_in(
+        &mut self,
+        reference: &BoundReference,
+        _literals: &FnvHashSet<Datum>,
+        _predicate: &BoundPredicate,
+    ) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+}
+
+/// A visitor to convert Iceberg bound predicates to Arrow predicates.
+struct PredicateConverter<'a> {
+    /// The leaf column indices used in the predicates.
+    pub columns: &'a Vec<usize>,
+    /// The projection mask for the Arrow predicates.
+    pub projection_mask: ProjectionMask,
+    /// The Parquet schema descriptor.
+    pub parquet_schema: &'a SchemaDescriptor,
+    /// The map between field id and leaf column index in Parquet schema.
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+impl PredicateConverter<'_> {
+    /// When visiting a bound reference, we return the projection mask for the leaf column
+    /// which is used to project the column in the record batch.
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<ProjectionMask> {
+        // The leaf column's index in Parquet schema.
+        let column_idx = self.column_map.get(&reference.field().id).ok_or_else(|| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!("Field id {} not found in schema", reference.field().id),
+            )
+        })?;
+
+        // Find the column index in projection mask.
+        let column_idx = self
+            .columns
+            .iter()
```

Comment:
Why not just `Vec::get` ? This iteration needs to go through all columns.

Reply:
Here we want to get the vector index of the Parquet leaf column index. For example, assume the vector is vector![1, 0, 3, 2], and the Parquet column index is 3, we need to get 2.

Reply:
Actually this looking up is not necessary. Removed it.